### ISSUE TITLE
OCIEnvCreate failed to create environment handle

### ISSUE THREAD

**Issue by jerome-peng** [laurenz/oracle_fdw#254]
I use the root user to install

    # yum install oracle-instantclient11.2-basic-11.2.0.4.0-1.x86_64.rpm 
    # yum install oracle-instantclient11.2-devel-11.2.0.4.0-1.x86_64.rpm 

I use the `report` user which is the owner of the PG server.
I can use sqlplus to login to oracle server as user `report`.

here is my error:

    report=> SELECT oracle_diag('server_ora_99');
    ERROR:  error connecting to Oracle: OCIEnvCreate failed to create environment handle
    DETAIL:  
    report=> select oracle_diag();
                                                 oracle_diag                                              
    ------------------------------------------------------------------------------------------------------
     oracle_fdw 2.0.0, PostgreSQL 9.5.12, Oracle client 11.2.0.4.0, ORACLE_HOME=/opt/oracle/instantclient
    (1 row)

**Comment by laurenz** [comment]
Oh no, the dreaded "OCIEnvCreate failed to create environment handle".

It means that *something* is wrong with your Oracle setup. Oracle does not tell you what. Often some environment variable.

As a first step, you can check all Oracle related environment variables you have set in the shell where sqlplus works. Perhaps some of them are not set in the environment of the PostgreSQL server.

Which naming method do you use to connect to Oracle? Can I see the connection string?

You might want to look through all the past issues that mention `OCIEnvCreate`, maybe you get a clue there.

Sometimes using `strace` to trace the PostgreSQL backend is helpful; it might show you where Oracle goes wrong.

**Comment by jerome-peng** [comment]
thanks a lot.

my problem is solved by reboot the server.

below is my environment variable.

LD_LIBRARY_PATH=/opt/oracle/instantclient:/home/jerome/work/pgsql/lib:/lib64:/usr/lib64:/usr/local/lib64:/lib:/usr/lib:/usr/local/lib
OCI_INC_DIR=/opt/oracle/instantclient/sdk/include
OCI_LIB_DIR=/opt/oracle/instantclient
ORACLE_HOME=/opt/oracle/instantclient

**Comment by kumakichi** [comment]
I came cross the same problem, and solved, with ubuntu 18.04.1(docker)

paste the steps here, may help somebody with the same problem:

1. find out how to start server 
  + start server with `service postgresql start` or `/etc/init.d/postgresql start`
  + find the command `ps ax | grep postgres`, you will get something like 
> /usr/lib/postgresql/10/bin/postgres -D /var/lib/postgresql/10/main -c config_file=/etc/postgresql/10/main/postgresql.conf

2. prepare
  + stop it firstly
      `service postgresql stop`
  + su(root user is not allowed to start server by default)
       `su - postgres`

3. add environment variables

```
export ORACLE_HOME=/opt/oracle/instantclient_12_2

export LD_LIBRARY_PATH=$ORACLE_HOME:/lib64:/usr/lib64:/usr/local/lib64:/lib:/usr/lib:/usr/local/lib
```

4. start server
run command got from step 1 , here, it is:
`/usr/lib/postgresql/10/bin/postgres -D /var/lib/postgresql/10/main -c config_file=/etc/postgresql/10/main/postgresql.conf`

then, connect server again, you will find this problem solved.

**Comment by laurenz** [comment]
For the record: On Linux, it is better to put Oracle on the library path by adding the directory to `/etc/ld.so.conf` and running `ldconfig`.
